### PR TITLE
feat(sequencer-client): implement client for querying and submitting txs to sequencer

### LINKED ISSUE #110: sequencer: client for constructing/submitting transactions

### PR DESCRIPTION
**Pull request by noot**
closes  #110

- implement a client crate for the sequencer node
- connects to the tendermint/cometbft endpoint, has functions for getting blocks, querying sequencer account state (balance, nonce), and submitting txs
- mock server for testing

as for the tendermint patch, see https://github.com/informalsystems/tendermint-rs/pull/1322 for details

### CODE REVIEW

File: crates/astria-sequencer-client/src/client.rs [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+                height,
+                false,
+            )
+            .await
+            .wrap_err("failed to call abci_query")?;
+
+        let balance = QueryResponse::try_from_slice(&response.value)
+            .wrap_err("failed to deserialize balance bytes")?;
+
+        if let QueryResponse::BalanceResponse(balance) = balance {
```
Comment (SuperFluffy) [inline crates/astria-sequencer-client/src/client.rs:105]: Nit: replace this `if let` with a `let QueryReponse::BalanceResponse(balance) = balance else { bail!(...) }`

File: crates/astria-sequencer-client/src/client.rs [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+            Ok(balance)
+        } else {
+            bail!("received invalid response from server: {:?}", &response);
```
Comment (SuperFluffy) [inline crates/astria-sequencer-client/src/client.rs:108]: Please also note which response variant you expected

File: crates/astria-sequencer-client/src/client.rs [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+            .abci_query(
+                Some(format!("accounts/balance/{}", &address.to_string())),
+                vec![],
+                height,
+                false,
+            )
+            .await
+            .wrap_err("failed to call abci_query")?;
+
+        let balance = QueryResponse::try_from_slice(&response.value)
```
Comment (SuperFluffy) [inline crates/astria-sequencer-client/src/client.rs:102]: Nit: I was trying to understand how/where Borsht is used and found this after going from the doc to here.

It's more verbose, but is `<QueryReponse as BorshtDeserialize>::try_from_slice` better?

File: crates/astria-sequencer-client/src/client.rs [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+    /// - If the response from the server is invalid.
+    pub async fn get_nonce(
+        &self,
+        address: &Address,
+        height: Option<Height>,
+    ) -> eyre::Result<Nonce> {
+        let response = self
+            .client
+            .abci_query(
+                Some(format!("accounts/nonce/{}", &address.to_string())),
```
Comment (SuperFluffy) [inline crates/astria-sequencer-client/src/client.rs:127]: Isn't adress implementing Display? If is you should throw it directly into the format!

File: crates/astria-sequencer-client/src/client.rs [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+                height,
+                false,
+            )
+            .await
+            .wrap_err("failed to call abci_query")?;
+
+        let nonce = QueryResponse::try_from_slice(&response.value)
+            .wrap_err("failed to deserialize balance bytes")?;
+
+        if let QueryResponse::NonceResponse(nonce) = nonce {
```
Comment (SuperFluffy) [inline crates/astria-sequencer-client/src/client.rs:138]: let ... else is a bit cleaner

File: crates/astria-sequencer-client/src/client.rs [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+            Ok(nonce)
+        } else {
+            bail!("received invalid response from server: {:?}", &response);
```
Comment (SuperFluffy) [inline crates/astria-sequencer-client/src/client.rs:141]: Please report which response you were expecting

File: crates/astria-sequencer-client/src/client.rs [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+        }
+    }
+
+    /// Submits the given transaction to the Sequencer node.
```
Comment (SuperFluffy) [inline crates/astria-sequencer-client/src/client.rs:145]: Empty line in the markdown here, i.e.

```markdown
/// Summary
///
/// More info
```

File: crates/astria-sequencer-client/src/client.rs [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+    /// This method blocks until the transaction is checked, but not until it's committed.
+    /// It returns the results of `CheckTx`.
```
Comment (SuperFluffy) [inline crates/astria-sequencer-client/src/client.rs:147]: And also here

**Review by SuperFluffy — APPROVED** [review]
Nice mocks and good error context. 🥰

Left a few nits but nothing major.

Making the contents of sequencer public is a bit unfortunate but I understand why it's necessary right now.

We should move the public interface of the sequencer into a dedicated crate.

**Review by joroshiba — APPROVED** [review]

File: Cargo.toml [hunk 9]
```diff
@@ -50,3 +52,11 @@ tracing = "0.1"
 tracing-subscriber = "0.3"
 uuid = "1.3.1"
 which = "4.4.0"
+
+[patch.crates-io]
+# fix for astria-sequencer-client 
+# https://github.com/informalsystems/tendermint-rs/compare/v0.32.0...astriaorg:tendermint-rs:noot/v0.32.0-fix?expand=1
```
[the remark this answers is not in the document]
Comment (noot) [reply Cargo.toml:58]: the patch was merged :D https://github.com/informalsystems/tendermint-rs/commit/9854a66205cdbea720f016061a9d427bd8162b12 
however, tendermint-rs main contains other changes for tendermint 0.38 and removes some of the ABCI methods, and hasn't been tagged as a release yet. will open an issue to update this at a later date

File: crates/astria-sequencer-client/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+[dependencies]
+astria-sequencer = { path = "../astria-sequencer" }
+
+tendermint-rpc = { version = "0.32", features = ["http-client"]}
+
+borsh = { workspace = true }
+ed25519-consensus = { workspace = true }
+eyre = { workspace = true }
+hex = { workspace = true }
+reqwest = { workspace = true, features = ["json"] }
```
[the remark this answers is not in the document]
Comment (noot) [reply crates/astria-sequencer-client/Cargo.toml:15]: this was unused actually, removed!

File: crates/astria-sequencer-client/src/client.rs [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+        query::Response as QueryResponse,
+        types::{
+            Address,
+            Balance,
+            Nonce,
+        },
+    },
+    transaction::Signed,
+};
+use borsh::BorshDeserialize;
```
[the remark this answers is not in the document]
Comment (noot) [reply crates/astria-sequencer-client/src/client.rs:12]: yeah, not totally sure, agree that is it annoying but it works for storing stuff to the state. we should probably compare serialization methods for storing things to disk at some point

**Pull request merged**